server/src/main/java/org/elasticsearch/common/util/concurrent/EsRejectedExecutionHandler.java

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.util.concurrent;
+
+import org.elasticsearch.common.metrics.CounterMetric;
+
+import java.util.concurrent.RejectedExecutionHandler;
+import java.util.concurrent.ThreadPoolExecutor;
+
+public abstract class EsRejectedExecutionHandler implements RejectedExecutionHandler {
+
+    private final CounterMetric rejected = new CounterMetric();
+
+    /**
+     * The number of rejected executions.
+     */
+    public long rejected() {
+        return rejected.count();
+    }
+
+    protected void incrementRejections() {
+        rejected.inc();
+    }
+
+    protected final EsRejectedExecutionException newRejectedException(Runnable r, ThreadPoolExecutor executor, boolean isExecutorShutdown) {
+        return new EsRejectedExecutionException("rejected execution of " + r + " on " + executor, isExecutorShutdown);
```

Comment:
I wonder if we should add info to the exception message about it being shutdown when `isExecutorShutdown=true`? I think that logging the exception will not show the flag otherwise.

Reply:
I added `(shutdown)` in the message for that.